src/server.rs

```diff
@@ -413,52 +410,59 @@ impl ServerState {
                     Some(player) => {
-
-                        // MESSAGE UPDATING
-                        match player.last_acked_msg_seq_num {
+                        let unsent_messages: Vec<BroadcastChatMessage>;
+                        // Only send what a player has not yet seen
+                        let raw_unsent_messages = match player.last_acked_msg_seq_num {
                             Some(last_acked_msg_seq_num) => {
-                                if !slot.messages.is_empty() {
-                                    let mut message_iter = slot.messages.iter();
-                                    let message = message_iter.next().unwrap(); // From front to back
-                                    let mut amount_to_consume = 0; // Skip over these messages since we've already acked them
-
-                                    if last_acked_msg_seq_num - message.seq_num > 0{
-                                        amount_to_consume = (last_acked_msg_seq_num - message.seq_num) % (MAX_AGE_CHAT_MESSAGES as u64);
-                                    }
-
-                                    let mut message_iter = message_iter.skip(amount_to_consume as usize);
-
-                                    while let Some(message) = message_iter.next() {
-                                        if message.player_id != player.player_id {
-                                            unsent_messages.push(BroadcastChatMessage{
-                                                chat_seq:    Some(message.seq_num),
-                                                player_name: message.player_name.clone(), // XXX is clone really the best way to do this?
-                                                message:     message.message.clone()
-                                            });
-                                        }
-                                    }
+                                // Unwrap()'s okay because we know its non-empty
+                                let mut message_iter = slot.messages.iter();
+
+                                let newest_msg = message_iter.clone().last().unwrap();
+                                // Player is caught up
+                                if last_acked_msg_seq_num == newest_msg.seq_num {
+                                    continue;
                                 }
+
+                                let oldest_msg = message_iter.clone().next().unwrap();
+                                // Skip over these messages since we've already acked them
+                                let amount_to_consume: u64 =
+                                    if last_acked_msg_seq_num > oldest_msg.seq_num {
+                                        ((last_acked_msg_seq_num - oldest_msg.seq_num) + 1) % (MAX_NUM_CHAT_MESSAGES as u64)
+                                    } else if last_acked_msg_seq_num < oldest_msg.seq_num {
+                                        // Sequence number has wrapped
```

Comment:
We should test this! Initialize sequence number to just less than max u64 and see what happens. :boom: